lib/logging/src/main/java/org/sagebionetworks/profiler/Frame.java

```diff
@@ -1,223 +1,124 @@
-	private List<Frame> children = null;
+	private static final String INDENT_STRING = "----";
+	private static final String DURATION_FORMAT_STRING = "HH:mm:ss.S";
 
-	public Frame() {
+	public Frame(String name) {
+		this.name = name;
+		this.elapsedTimeStatistics = new IntervalStatistics();
+		this.children = new LinkedHashMap<>();
 	}
 
-	public Frame(long startTime, String name) {
-		this.startTime = startTime;
-		this.name = name;
+	public Frame addChildFrameIfAbsent(String methodName){
+		Frame childFrame = children.get(methodName);
+		if (childFrame == null){
+			childFrame = new Frame(methodName);
+			children.put(methodName, childFrame);
+		}
+		return childFrame;
 	}
 
 	public String toString() {
 		// Output a single line if there are no children.
 		long id = Thread.currentThread().getId();
 		StringBuilder buffer = new StringBuilder();
-		buffer.append(String.format(
-				"%n[%3$d] ELAPSE: %1$tM:%1$tS:%1$tL METHOD: %2$s", elapse,
-				name, id));
-		if (children == null) {
-			return buffer.toString();
-		} else {
-			printChildren(buffer, id, "----", System.currentTimeMillis());
-			// Now print the children
-			return buffer.toString();
+		buffer.append(String.format("%n[%d] ELAPSE: ", id));
+		appendStatisticsToStringBuilder(buffer, this);
+		if (children != null) {
+			printChildren(buffer, id, 1);
 		}
+		return buffer.toString();
 	}
 
-	public void printChildren(StringBuilder buffer, long id, String level,
-			long time) {
-		if (children != null) {
-			for (int i = 0; i < children.size(); i++) {
-				Frame child = children.get(i);
-				buffer.append(String
-						.format("%n[%5$d] ELAPSE: %2$s%1$tM:%1$tS:%1$tL METHOD: %3$s",
-								child.elapse, level, child.name, time, id));
-				if (child.children != null) {
-					child.printChildren(buffer, id, level + "----", time);
-				}
-			}
+	public void printChildren(StringBuilder buffer, long id, int level) {
+		String indentString = StringUtils.repeat(INDENT_STRING, level);
+		for (Frame child: children.values()) {
+			buffer.append(String.format("%n[%d] ELAPSE: %s ", id, indentString));
+			appendStatisticsToStringBuilder(buffer, child);
+			child.printChildren(buffer, id, level + 1);
 		}
 	}
 
-	public void setEnd(long end) {
-		this.elapse = (end - startTime)/1000000;
+	private static void appendStatisticsToStringBuilder(StringBuilder buffer, Frame frame){
+		buffer.append( String.format("%s METHOD: %s", DurationFormatUtils.formatDuration(frame.getTotalTimeMilis(), DURATION_FORMAT_STRING), frame.getName()));
+		if (frame.getCallsCount() > 1){
+			buffer.append(String.format(" (Count: %d Average: %s, Min: %s, Max: %s)",
+										frame.getCallsCount(),
+										DurationFormatUtils.formatDuration(frame.getAverageTimeMilis(), DURATION_FORMAT_STRING),
+										DurationFormatUtils.formatDuration(frame.getMinTimeMilis(), DURATION_FORMAT_STRING),
+										DurationFormatUtils.formatDuration(frame.getMaxTimeMilis(), DURATION_FORMAT_STRING)));
+		}
 	}

```

Comment:
Looking for a test of a full format that includes hours, mins, secs, mils.
Also, test for appendStatisticsBuilder() for case with and without counts.